Create an advanced API Blueprint tutorial

danielgtaylor opened the pull request:
This (draft) creates a new advanced API Blueprint tutorial covering JSON Schema, request and response attributes, data structures and relation types.

I'm pretty unsure of the relation types section, so please have a look and give me any feedback. Once this is ready, I'm thinking of making available a final full example `.apib` file with these features and linking the tutorial to the advanced tutorial somehow.

Feedback will be much appreciated!

pksunkara commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,211 @@
+                    "choice": "Swift",
+                    "url": "/questions/1/choices/1",
+                    "votes": 2048
+                  }, {
```
Can we just have 2 choices here? We don't need as many. We should try to make the Tutorials as small as possible

danielgtaylor replied:
This was copied directly from the first tutorial, but I can make it have fewer choices if it helps.

pksunkara commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,211 @@
+                + (object)
+                    + choice: 'Javascript' (required)
+                    + url: '/questions/1/choices/1' (required)
+                    + votes: 2048 (number, required)
```
There is only 1 choice depicted here. Shouldn't a question have more than 1 choice?

danielgtaylor replied:
It is an array of choices with a single sample. It's not entirely clear to me how to add more samples when you have an array of objects. Do you know how?

pksunkara commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,211 @@
++ Relation: question_update
+
+### Delete a Question [DELETE]
++ Relation: question_delete
```
Do we need the `question` prefixes? It's not like there is a rule saying relations should be unique in the whole API Blueprint. What do you say @kylef, @zdne?

kylef replied:
I've called this relation `delete` in both the polls API (when you use hypermedia content types), and in the [polls-app](https://github.com/apiaryio/polls-app/blob/master/apiary.apib#L29)), we should certainly align these uses.

Is there any reason why you think `questions_delete` would be better over `delete`. I personally prefer the shorter variant so that a client can understand the semantic meaning of `delete` shared across multiple resources instead of a one-off relation which differs per resource.

pksunkara replied:
What would do if there are 2 delete methods in the API? 1 for questions, 1 for choices.

danielgtaylor replied:
@kylef I will rename them to match the polls app, but how do you link between resources in your approach?

pksunkara commented:
Looks good. Haven't checked closely for grammar issues.

kylef commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,211 @@
+
+    + Schema
+
+            {
```
This tutorial produces a large JSON Schema, I wonder if we would be better off describing a simpler structure, perhaps the structure used on creating a question instead:

```apib
+ Request (application/json)
    + Body

            {
              "question": "Favourite language?"
              "choices": [ "Swift", "Objective-C" ]
            }

    + Schema

            {
              "$schema": "http://json-schema.org/draft-04/schema#",
              "type": "object",
              "properties": {
                "question": { "type": "string }
                "choices": {
                  "type": "array",
                  "items": { "type": "string" },
                  "minItems": 2
                }
              }
            }
```

pksunkara replied:
I agree with @kylef

kylef commented on Advanced Tutorial.md:
````diff
@@ -0,0 +1,211 @@
+Action request and response bodies can have associated schemas that describe the allowed structure of the body content. JSON bodies are typically described with [JSON Schema](http://json-schema.org/). Given a simple JSON response body we can describe the structure of the response with JSON Schema in a `+ Schema` section:
+
+```apib
+## Random Numbers [/random]
````
I think it would be nice to have a tutorial that keeps the same domain along all of the examples so it's easier to follow. Instead of jumping from a random number generator to the polls app, we could make both examples related to the polls app.

kylef commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,211 @@
+
+## Data Structures
+
+Once you start using MSON, you may find yourself wanting to reuse certain commonly used or nested data structure components. This is possible with the `## Data Structures` section. Attributes sections can then reference the data structures defined in the Data Structures section by name.
```
I think it would be worth pointing out that you can reference both data structures and other resources by name.

kylef commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,211 @@
++ choices (array[Choice], required)
+
+### Choice
++ choice: 'Javascript' (required)
```
Above you've placed strings directly and here wrapped them in quotes. Can we remove the quotes here to make this consistent with the other example `question` in Question above.

kylef commented on Advanced Tutorial.md:
````diff
@@ -0,0 +1,211 @@
++ votes: 2048 (number, required)
+```
+
+## Relation Types
````
Seems to be a bug in GitHub's highlighting here when you view the diff. I wonder if the bug is caused from our `apib` lexer not completing correctly from the above code block? /cc @danielgtaylor

pksunkara replied:
I think it's a markdown bug. It will be fixed if we add an empty line before the ending of above snippet with 3 backticks

kylef commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,211 @@
+
+## Conclusion
+
+This tutorial has covered some advanced API Blueprint topics. For more in-depth information and other advanced topics, please see the [API Blueprint Specification](https://github.com/apiaryio/api-blueprint/blob/master/API%20Blueprint%20Specification.md).
```
I think it would be nice to construct the full version of the above example(s) and link to it in the conclusion. So a user can view a complete blueprint the tutorial builds after the tutorial.

kylef commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,211 @@
++ Relation: question
+
+### Update a Question [PUT]
++ Relation: question_update
```
The Polls API doesn't allow you to update a question and I don't think it makes sense given our polls domain. Questions shouldn't be modifiable since users can vote on them.

I'd suggest we drop this action from the example.

kylef commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,211 @@
+
+## Relation Types
+
+Actions in a blueprint can define a semantic domain-specific meaning by defining a relation type using a `+ Relation` section. This means that an action can have a specific meaning regardless of its URI or name, and allows clients to be built based on the domain of the API rather than specific URIs.
```
I'd love to see a link to the RFC defining [web linking](https://tools.ietf.org/html/rfc5988) either here or in the conclusion for this section of the tutorial. This allows a user to go off and read more.

kylef commented on Advanced Tutorial.md:
````diff
@@ -0,0 +1,211 @@
+```apib
+## Question [/question/{id}]
+### View a Question Detail [GET]
++ Relation: question
````
I think a relation to the same resource should be [`self`](http://www.iana.org/assignments/link-relations/link-relations.xhtml).

The [polls app](https://github.com/apiaryio/polls-app/blob/master/apiary.apib#L60) is also incorrect, this also needs to be changed but depends on [Draft API Blueprint RFC for affordances](https://github.com/apiaryio/api-blueprint-rfcs/pull/2).

danielgtaylor commented:
@kylef @pksunkara I think I have addressed all of the feedback. Please take another look!

I'll also look into the highlighting bugs soon, but that will be a different task.

pksunkara commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,119 @@
+
+Action request and response bodies can have associated schemas that describe the allowed structure of the body content. JSON bodies are typically described with [JSON Schema](http://json-schema.org/). Given a simple JSON response body we can describe the structure of the response with JSON Schema in a `+ Schema` section.
+
+The schema can describe the type of each member, which members are required, default values, and support a number of other advanced features. Below is an example, taken from the API Blueprint tutorial blueprint:
```
The below example is not taken from API Blueprint tutorial.

pksunkara commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,119 @@
+
+MSON can be added to resources, actions, and individual requests or responses via an `+ Attributes` section.
+
+The polls API question resource can be modeled using MSON:
```
The below example is not a polls API question resource.

danielgtaylor commented:
@pksunkara addressed feedback. Thanks!

pksunkara commented on Advanced Tutorial.md:
```diff
@@ -0,0 +1,123 @@
+
+Once you start using MSON, you may find yourself wanting to reuse certain commonly used or nested data structure components. This is possible with the `## Data Structures` section. Attributes sections can then reference the data structures defined in the Data Structures or other resource sections by name.
+
+For example, using the polls API question resource, we can split out the `Question` and `Choice` objects:
```
Isn't `List All Questions` part of `Question Collection` resource?

pksunkara commented on Advanced Tutorial.md:
````diff
@@ -0,0 +1,123 @@
++ choice: Javascript (required)
++ url: /questions/1/choices/1 (required)
++ votes: 2048 (number, required)
+```
````
Add an empty line before this line and the syntax highlighting should be fixed.

kylef commented on Advanced Tutorial.md:
````diff
@@ -0,0 +1,123 @@
+
+```
+
+When the above blueprint is parsed it will have a JSON body and JSON Schema example generated for it from the MSON attributes.
````
Perhaps it might be worth noting a user can override the schema, and the generated schema won't be the same as above as it won't include `minItems`.

danielgtaylor commented:
More feedback addressed. I can't seem to fix the syntax highlighting - it will need to be done along with a few other minor fixes in the syntax highlighter package. Will try and get to that soon, as it's quite annoying.

zdne commented:
👍
